---
title_full: Template for preparing a submission to Bioinformatics using RMarkdown
title_short: Bioinformatics Rmd Template

# One of the following below. See https://academic.oup.com/bioinformatics/pages/instructions_for_authors#Types%20of%20Manuscript
# one of `Original Paper`, `Discovery Note`, `Applications Note`, `Review`, `Letter to the Editor`, `Conference Paper`    
manuscript_type: Application Note

# change accordingly
pubyear: '`r format(Sys.Date(), "%Y")`'

# One of the following below. See https://academic.oup.com/bioinformatics/pages/instructions_for_authors#Detailed%20Scope%20Guidelines
# Genome analysis
# Sequence analysis
# Phylogenetics
# Structural bioinformatics
# Gene expression
# Genetic and population analysis
# Systems biology
# Data and text mining
# Databases and ontologies
# Bioimage informatics
subject_section:  Metagenomic analysis

access_date:
  - day: 28
    month: 05
    year: 2025
      
author:
  - name: Yamila Timmer
    affiliation: "1"
  - name: Floris Menninga
    affiliation: "2"
  - name: Jarno Jacob Duiker
    affiliation: "3"
          
address:
  - code: 1
    address: Hanze university of aplied sciences, Department, Street, Groningen, Groningen, Zip
  - code: 2
    address: Hanze university of aplied sciences, Street, Groningen, Groningen, Zip
 
# Full email address to be given, preferably an institution email address.  
corresponding_author_text: "To whom correspondence should be addressed. E-mail: bob\\@email.com"
corresponding_author_email: "bob\\@email.com"
  
motivation: |
  This section should specifically state the scientific question within the context of the field of study.
  
results: |
  This section should summarize the scientific advance or novel results of the study, and its impact on computational biology.

availability: |
  This section should state software availability if the paper focuses mainly on software development or on the implementation of an algorithm. Examples are: 'Freely available on the web at XXX.' Website implemented in Perl, MySQL and Apache, with all major browsers supported'; or 'Source code and binaries freely available for download at URL, implemented in C++ and supported on linux and MS Windows'. The complete address (URL) should be given. If the manuscript describes new software tools or the implementation of novel algorithms the software must be freely available to non-commercial users. Authors must also ensure that the software is available for a full TWO YEARS following publication. The editors of Bioinformatics encourage authors to make their source code available and, if possible, to provide access through an open source license.

supplementary_info: |
  Supplementary data are available at Bioinformatics Online.

output:
  rticles::bioinformatics_article:
    citation_package: natbib
    
bibliography: bibliography.bib
---

# Introduction

Cite others using bracket notation [@pepe2003statistical]. Can also cite with @zou2005regularization.

<!-- naked refs don't work because of the \nolinkurl problems -->

<!-- Or refer to specific people e.g. -@Yoo03 like so. -->

Instructions for authors are available [online](http://www.oxfordjournals.org/our_journals/bioinformatics/for_authors/general.html).

Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

**Abstract**

This study employs metagenomic tools to analyze wastewater from the Afac filtering lagoon in Kenya, which is used by an agrochemical company to treat river water for production processes before release back into the river. The goal is to evaluate the lagoon’s effectiveness in removing hazardous microorganisms and assess potential ecological impacts on downstream ecosystems and communities.

Sequencing data from the MinION platform underwent rigorous quality assessment and trimming using *Fastplong* [@fastplong][ @Chen2018], followed by taxonomic classification with *Kraken 2* [@Wood2019]. Results were visualized interactively using *KronaTools* [@Ondov2011] and *Pavian* (Breitwieser & Salzberg, 2020), enabling detailed taxonomic profiling. Microbial functional pathways were analyzed with *HUMAnN 3.0* (Beghini et al., 2021), while alpha- and beta-diversity metrics were decided using a krona-plugin python script

Our analysis identifies microorganisms persisting through the lagoon’s treatment stages, providing insights into filtration efficacy and risks of downstream pollution. The findings will serve as actionable recommendations to optimize the lagoon system, mitigate ecological harm, and safeguard river health for adjacent communities.

**Introduction**

Country’s in northern Africa, The Middle East and Singapore, Maldives and Australia face water shortiges. Making water quite a valuable resource, because of this there Country’s and regions have practiced reclaiming waste water from factory’s and sewage. Instead of wasting this water it could help against the big shortages and help the infrastructure stay up and active. \
Reclaimed water is used in various places, most known is the landscape irrigation to maintain green living spaces and for agricultural irrigation to produce food. (Hong et al., 2020) Reclaimed water is also key for important infrastructures like cooling towers that serve electrical power plants. All these things mean that reclaiming this waste water is key to some country's daily operation. \
\
Reclaiming waste water does come with some risks. The water used in the factory's may contain chemicals or microorganisms that could have a negative influence on the health of people or the ecosystem.(Chen et al., 2013) Many of these factory's use filters to take these chemicals and or microorganisms out of the water so it can return safely into the river or other water source, however not every factory uses the same filter due to costs and or the infrastructure to make these filters is not available. Thorough filtering is key in reclaiming waste water, because many biological contaminants are disseminated through water, and their occurrence has potential detrimental impacts on public and environmental health. (Hong et al., 2020)

Metagenomics is a way to look and identify microorganisms by their sequences, this project uses metagenomics for the identification of a multitude of microorganisms found in waste water from the lagoon of an AFAC factory in Kenya. Building a pipeline with tools for the metagenomic analysis of this waste water gives the factory the chance to reuse this pipeline so it can make progress in improving it’s lagoon filtering property’s.

The factory takes water from a nearby river, uses it in it's production then filters it through the lagoon, the lagoon is chosen because it is a cheap way to let nature filter out the chemicals. Often in remote places in africa it is quite hard to recieve the correct materials and often it is also quite expensive. After the water has finished going through the lagoon it returns the water to the previously mentioned river. This river is not only used by the factory but is also the main water source for villages down stream. This means if this water is contaminated with pathogens or other dangerous substances (heavy metals, chemicals, etc) it will have a significant impact on the villagers but also the ecosystem that the river contains. The ecosystem of the river is also a food source for the villagers meaning there is double impact if there are problems within the river.\
\
For the analysis of the waste water samples have been taken out of this filter in different stages. Analysis of these different stages of the lagoon will give insight on the effectiveness of the filtering done by the lagoon and what survives after the filtering in the lagoon which will return back to the river . These samples have shown significant bacterial growth in lab tests conducted previously, these lab tests have been very insightful in showing that even in different conditions there are still bacteria to be found from this waste water. Due to these lab tests this project has been started to look at the sequence of the grown microorganism and identify them. These sequences are coming from a minION high throughput sequencer. Using meta genomic tools it is possible to see what kind of microorganisms have grown in the sample water. The next step is to show how dangerous each microorganism is with the use of visualization, this will give more insight on how to improve the lagoon's filtering property's.

# Approach

Here is how to include math equations in the document (bounded by `$$`):

**Shannon index for alpha diversity**

$$
\begin{aligned}
H=−∑[(pi​)×log(pi​)]
\end{aligned}
$$

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

```{=html}
<!--
problems with longtable and 2-column format. see pnas_template and https://github.com/jgm/pandoc/issues/1023

  Species                    CBS     CV     G3
  ----------------------- ------ ------ ------
  1\. Acetaldehyde           0.0    0.0    0.0
  2\. Vinyl alcohol          9.1    9.6   13.5
  3\. Hydroxyethylidene     50.8   51.2   54.0

  : Comparison of the fitted potential energy surfaces and ab initio
  benchmark electronic energy calculations
-->
```
Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

```{r figure, echo = FALSE, fig.width = 3.5, fig.cap = 'Figure from an Rmd chunk.'}
plot(rnorm(10000))
```

## 2. Materials and methods

#### 2.1 overall approach

The sequencing data obtained from the MinION platform underwent a quality assessment process. Initially, the sequences were evaluated for quality using `Fastplong` @fastplong @Chen2018. After this, `Fastplong` also trimmed the data according to the generated reports. Following this, a secondary quality assessment was conducted using `Fastplong` to ensure improved data quality. The refined dataset was then subjected to taxonomic classification through `Kraken2` @Wood2019, a computational tool for microbial classification. The results generated by `Kraken2` were then visualised using `KronaTools` (Ondov et al., 2011) and `Pavian` (Breitwieser & Salzberg, 2020), facilitating an interactive and intuitive representation of the taxonomic distribution. Other tools like `HUMAnN 3.0` (Beghini et al., 2021) were used for profiling the abundance of microbial metabolic pathways and other molecular functions from metagenomics data (Beghini et al., 2021). Furthermore, the outcomes obtained from the `Kraken2` analysis were employed to identify the microorganisms present in the lagoon water, this allowed comparing the different lagoon stages and their microbiomes. To get more insight into the data using 2 kraken2 plugins that will calculate the alpha- and beta-diversity, a tool that made a Shannon index for alpha diversity, this index showed the species diversity in each sample. This index also showed the distribution of the species in the samples. With all this information, an analysis for each stage of the lagoon has been made.

#### 2.2 Data collection

**QIAamp DNA Microbiome Kit (50) - 51704** - What is this kit used for: This is a purification and enrichment kit of bacterial microbiome DNA from swabs (and body fluids). Effective depletion of host DNA during the purification process maximises bacterial DNA coverage in NGS analysis and allows for 16rDNA-based microbiome analysis and whole metagenome shotgun sequencing studies

**Procedure**

The kit employs spin column technology with a specialized protocol to enrich bacterial microbiome DNA while minimizing host DNA contamination. First, host cells are gently lysed, and their released DNA is enzymatically degraded. Next, bacterial cells are disrupted using optimized mechanical and chemical lysis. The bacterial DNA is then selectively bound to a silica membrane, purified through washes, and finally eluted for analysis.

This method ensures efficient isolation of bacterial DNA from complex samples, reducing host DNA interference.

### 2.3 The metagenomics pipeline

#### 2.3.1 **Data Preprocessing and Quality Control**

Raw sequencing data has been received in the FASTQ format, this data was preprocessed using `fastplong` . A specialized version of the fastp tool optimised for Nanopore long-read data, adapter and low quality reads (Phred score \< 20) were trimmed, and reads shorter than 50 bp were no longer used to ensure the high-quality and keep the data integrity high for the downstream analysis. After this first trim, the data was assessed to see if the data quality had improved again using `fastplong`. It was concluded that the data did not need more trimming, allowing the next step to proceed.

#### 2.3.2 Identifying microorganisms using taxonomic classification

The processed FASTQ files were run through the pipeline's next step, `Kraken2`. `Kraken2` was used to get the taxonomic classification of each found microorganism. This was done by matching k-mer sequences (nucleotide sequences), giving `Kraken2` high accuracy. The output file has a certain structure with some important things like the taxonomy ID, whether a sequence was classified or not and a list containing the lowest common ancestor. The identities of the present microorganisms were revealed by their taxonomy ID, however, reading each ID and searching these up would be quite inefficient. This will be solved with the visualisation step, which takes these IDs and visualises them.

#### 2.3.3 Visualisation of kraken2 results

To visualise what has been found from the Taxonomic classification of the `Kraken2` tool, we used a multitude of different tools, `KronaTools` was used to create krona plot, a krona plot shows all taxonomic levels based on the NCBI taxonomy starting from superkingdom to family level and the associated abundances based on the number of identified spectra summed over all BGPs (Bacterial Gene Product). This was used in the paper to show what had been found in an easy to interpret way. `Pavian` was used to create an interactive browser application that analysed and visualised metagenomics classification results from classifiers such as `Kraken2`. `Pavian` also provided an alignment viewer to validate matches to a particular genome. Specific visualisations that could be found in the browser application are: a taxonomic classification chart, a Sankey diagram, and the alignment viewer.

#### 2.3.4 Other visualisations

`HUMAnN 3.0` was used to profile the abundance of microbial metabolic pathways and other molecular functions from the data. This showed the metabolic potential of the microbial community that was present in the lagoon samples. This helped us answer the question of “What are the microorganisms in the lagoon doing or capable of doing?”. `QIIME2`, a tool that made a Shannon index for alpha diversity, this index showed the species diversity in each sample. This index also showed the distribution of the species in the samples. After it was clear what the distribution was of these microorganisms, we could build a microbiome structure for each step in the lagoon. These structures have been used to answer the questions about the lagoon, “*Not clear yet*”

# Discussion

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

# Conclusion

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

# Acknowledgements {.unnumbered}

These should be included at the end of the text and not in footnotes. Please ensure you acknowledge all sources of funding, see funding section below.

Details of all funding sources for the work in question should be given in a separate section entitled 'Funding'. This should appear before the 'Acknowledgements' section.

# Funding {.unnumbered}

The following rules should be followed:

-   The sentence should begin: ‘This work was supported by …’ -
-   The full official funding agency name should be given, i.e. ‘National Institutes of Health’, not ‘NIH’ (full RIN-approved list of UK funding agencies)
-   Grant numbers should be given in brackets as follows: ‘[grant number xxxx]’
-   Multiple grant numbers should be separated by a comma as follows: ‘[grant numbers xxxx, yyyy]’
-   Agencies should be separated by a semi-colon (plus ‘and’ before the last funding agency)
-   Where individuals need to be specified for certain sources of funding the following text should be added after the relevant agency or grant number 'to [author initials]'.

An example is given here: ‘This work was supported by the National Institutes of Health [AA123456 to C.S., BB765432 to M.H.]; and the Alcohol & Education Research Council [hfygr667789].’

Oxford Journals will deposit all NIH-funded articles in PubMed Central. See Depositing articles in repositories – information for authors for details. Authors must ensure that manuscripts are clearly indicated as NIH-funded using the guidelines above.
